runtime/queries/metricsview.go

```diff
@@ -0,0 +1,312 @@
+	whereClause := ""
+	var args []any
+
+	if filter != nil && filter.Include != nil {
+		clause, clauseArgs, err := buildFilterClauseForConditions(filter.Include, false)
+		if err != nil {
+			return "", nil, err
+		}
+		whereClause += clause
+		args = append(args, clauseArgs...)
+	}
+
+	if filter != nil && filter.Exclude != nil {
+		clause, clauseArgs, err := buildFilterClauseForConditions(filter.Exclude, true)
+		if err != nil {
+			return "", nil, err
+		}
+		whereClause += clause
+		args = append(args, clauseArgs...)
+	}
+
+	return whereClause, args, nil
+}
+
+func buildFilterClauseForConditions(conds []*runtimev1.MetricsViewFilter_Cond, exclude bool) (string, []any, error) {
+	clause := ""
+	var args []any
+
+	for _, cond := range conds {
+		condClause, condArgs, err := buildFilterClauseForCondition(cond, exclude)
+		if err != nil {
+			return "", nil, fmt.Errorf("filter error: %s", err.Error())
+		}
+		if condClause == "" {
+			continue
+		}
+		clause += condClause
+		args = append(args, condArgs...)
+	}
+
+	return clause, args, nil
+}
+
+func buildFilterClauseForCondition(cond *runtimev1.MetricsViewFilter_Cond, exclude bool) (string, []any, error) {
+	var clauses []string
+	var args []any
+
+	var operatorPrefix string
+	var conditionJoiner string
+	if exclude {
+		operatorPrefix = " NOT "
+		conditionJoiner = ") AND ("
+	} else {
+		operatorPrefix = ""
+		conditionJoiner = " OR "
+	}
+
+	if len(cond.In) > 0 {
+		// null values should be added with IS NULL / IS NOT NULL
+		nullCount := 0
+		for _, val := range cond.In {
+			if _, ok := val.Kind.(*structpb.Value_NullValue); ok {
+				nullCount++
+				continue
+			}
+			arg, err := protobufValueToAny(val)
+			if err != nil {
+				return "", nil, fmt.Errorf("filter error: %s", err.Error())
+			}
+			args = append(args, arg)
+		}
+
+		questionMarks := strings.Join(repeatString("?", len(cond.In)-nullCount), ",")
+		// <dimension> (NOT) IN (?,?,...)
+		if questionMarks != "" {
+			clauses = append(clauses, fmt.Sprintf("%s %s IN (%s)", cond.Name, operatorPrefix, questionMarks))
+		}
+		if nullCount > 0 {
+			// <dimension> IS (NOT) NULL
+			clauses = append(clauses, fmt.Sprintf("%s IS %s NULL", cond.Name, operatorPrefix))
+		}
+	}
+
+	if len(cond.Like) > 0 {
+		for _, val := range cond.Like {
+			arg, err := protobufValueToAny(val)
+			if err != nil {
+				return "", nil, fmt.Errorf("filter error: %s", err.Error())
+			}
+			args = append(args, arg)
+			// <dimension> (NOT) ILIKE ?
+			clauses = append(clauses, fmt.Sprintf("%s %s ILIKE ?", cond.Name, operatorPrefix))
+		}
+	}
+
+	clause := ""
+	if len(clauses) > 0 {
+		clause = fmt.Sprintf(" AND (%s)", strings.Join(clauses, conditionJoiner))
+	}
+	return clause, args, nil
+}
+
+func repeatString(val string, n int) []string {
+	res := make([]string, n)
+	for i := 0; i < n; i++ {
+		res[i] = val
+	}
+	return res
+}
+
+func protobufValueToAny(val *structpb.Value) (any, error) {
```

Comment:
move to util ?

Reply:
Related to my comment about `valToPB`, this could become `pbutil.FromValue(v *structpb.Value) (any, error)`